model/jpa/src/main/java/org/keycloak/models/jpa/session/PersistentUserSessionEntity.java

```diff
@@ -32,16 +32,32 @@
         @NamedQuery(name="updateUserSessionLastSessionRefresh", query="update PersistentUserSessionEntity sess set lastSessionRefresh = :lastSessionRefresh where sess.realmId = :realmId" +
                 " AND sess.offline = :offline AND sess.userSessionId IN (:userSessionIds)"),
-        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline"),
+        @NamedQuery(name="findUserSessionsCount", query="select count(sess) from PersistentUserSessionEntity sess where sess.offline = :offline and sess.realmId = coalesce(:realmId, sess.realmId)"),
         @NamedQuery(name="findUserSessions", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
                 " AND (sess.createdOn > :lastCreatedOn OR (sess.createdOn = :lastCreatedOn AND sess.userSessionId > :lastSessionId))" +
-                " order by sess.createdOn,sess.userSessionId")
-
+                " order by sess.createdOn,sess.userSessionId"),
+        @NamedQuery(name="findUserSession", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
+                " AND sess.userSessionId = :userSessionId AND sess.realmId = :realmId"),
+        @NamedQuery(name="findUserSessionsByUserId", query="select sess from PersistentUserSessionEntity sess where sess.offline = :offline" +
+                " AND sess.realmId = :realmId AND sess.userId = :userId"),
+        @NamedQuery(name="findUserSessionsByClientId", query="SELECT sess FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
+                " ON sess.userSessionId = clientSess.userSessionId AND clientSess.clientId = :clientId WHERE sess.offline = :offline " +
+                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId AND clientSess.clientId = :clientId"),
+        @NamedQuery(name="findUserSessionsCountsByClientId", query="SELECT clientSess.clientId, count(clientSess) " +
+                " FROM PersistentUserSessionEntity sess INNER JOIN PersistentClientSessionEntity clientSess " +
+                " ON sess.userSessionId = clientSess.userSessionId " +
+                // find all available offline user-session for all or specific clients in a realm
+                " AND clientSess.clientId = coalesce(:clientId, clientSess.clientId) " +
+                " WHERE sess.offline = :offline " +
+                " AND sess.userSessionId = clientSess.userSessionId AND sess.realmId = :realmId " +
+                // find all available offline user-session for all or specific clients in a realm
+                " AND clientSess.clientId = coalesce(:clientId, clientSess.clientId) " +
+                " GROUP BY clientSess.clientId")
```

Comment:
Most of these queries resort to plain seq scan because of missing / inappropriate indices.

Reply:
The query `findUserSession` results in the following SQL:
```sql
--- PersistentUserSessionEntity.findUserSession

select persistent0_.OFFLINE_FLAG         as OFFLINE_1_41_,
       persistent0_.USER_SESSION_ID      as USER_SES2_41_,
       persistent0_.CREATED_ON           as CREATED_3_41_,
       persistent0_.DATA                 as DATA4_41_,
       persistent0_.LAST_SESSION_REFRESH as LAST_SES5_41_,
       persistent0_.REALM_ID             as REALM_ID6_41_,
       persistent0_.USER_ID              as USER_ID7_41_
from OFFLINE_USER_SESSION persistent0_
where persistent0_.OFFLINE_FLAG='1'
  and persistent0_.USER_SESSION_ID='400c53b3-1c21-40a5-8a40-727f1f772805'
  and created_on >0
  and persistent0_.REALM_ID='acme'
limit 1
```
by default the primary key index  is used:
![image](https://user-images.githubusercontent.com/314690/107161768-3aac0280-699f-11eb-94a5-cf8ea8a3d0f3.png)

with a more specific index:
```sql
create index idx_offline_uss_by_usersess on offline_user_session (offline_flag, realm_id, user_session_id);
```
the query is much quicker:
![image](https://user-images.githubusercontent.com/314690/107161717-e43ec400-699e-11eb-8f72-6ed0acf898fe.png)


The query `findUserSessionsByUserId` generates the following SQL:
```sql
--- PersistentUserSessionEntity.findUserSessionsByUserId

select persistent0_.OFFLINE_FLAG         as OFFLINE_1_41_,
       persistent0_.USER_SESSION_ID      as USER_SES2_41_,
       persistent0_.CREATED_ON           as CREATED_3_41_,
       persistent0_.DATA                 as DATA4_41_,
       persistent0_.LAST_SESSION_REFRESH as LAST_SES5_41_,
       persistent0_.REALM_ID             as REALM_ID6_41_,
       persistent0_.USER_ID              as USER_ID7_41_
from OFFLINE_USER_SESSION persistent0_
where persistent0_.OFFLINE_FLAG='1'
  and persistent0_.REALM_ID='acme'
  and persistent0_.USER_ID='87511de7-165c-4a34-891d-d0db60872644'
```
The uses a combined bitmap scan
![image](https://user-images.githubusercontent.com/314690/107161605-0d128980-699e-11eb-9824-50577937fb56.png)

If I create a more specific index:

```sql
create index idx_offline_uss_by_user on offline_user_session (offline_flag, realm_id, user_id);
```

Then the query is much quicker:
![image](https://user-images.githubusercontent.com/314690/107161624-2c111b80-699e-11eb-88a8-ee5f7e364a68.png)

Reply:
With this I'd say that the new queries are all covered by an index.

Reply:
I think the index definitions I choose are not optimal and need some additional tweaking.